### PR TITLE
Add attribute-based overloading

### PR DESCRIPTION
**Pull request by Baltoli**
This PR is the next step towards the changes suggested in https://github.com/runtimeverification/k/issues/3966; having added a new syntax for the `symbol(_)` attribute in #3978, we can now address the second use of `klabel(_)` (marking overloaded sets of symbols).

The changes in this PR are as follows:
* Add a new attribute `overload(_)` that behaves identically to the current semantics of `klabel(_)` with respect to the construction of the overload partial ordering, but that does not participate in the _naming_ of symbols.
* Add checks for the new attribute:
  * Disallow `klabel(_)` and `overload(_)` on the same production; overload sets must be converted fully to use the new mechanism.
  * Warn when an overload set of only one production is generated.
* Document the semantics of overloading in the user manual based on our conversations in #3966.
* Add a new integration test that checks a simple example of overloading works as expected to disambiguate parses.

Once this PR is merged, we can begin the long process of deprecating `klabel(_)` and `symbol`!

### CODE REVIEW

**Comment by Baltoli** [comment]
Because we emit a _different_ overload attribute to KORE; this is actually blocked on a change @Scott-Guest is working on to strip frontend attributes before generating KORE.

**Comment by tothtamas28** [comment]
The Haskell backend seems to depend on an existing `overload` attribute: https://github.com/runtimeverification/k/issues/3946#issuecomment-1930372019.

**Comment by Baltoli** [comment]
Indeed - that overload attribute is the _KORE_ overload attribute (which IIRC means something like "this axiom was generated from the overload partial ordering, and needs special treatment for unification" - a different semantics to the K attribute, which is something like "this symbol should participate in overload selection").

The plan is for @Scott-Guest to finish his work on #3983, which will conclude with a compiler pass that strips out attributes meaningful to K but not KORE before `ModuleToKore` is performed. This means we can keep the K `overload(_)` syntax in the frontend as well as the existing semantics of `overload` in KORE.

**Comment by tothtamas28** [comment]
> strips out attributes meaningful to K but not KORE before `ModuleToKore` is performed

There are steps in `ModuleToKORE` that depend on the overload lattice, will it still be reproducible after stripping the K `overload` attribute?

**Comment by Baltoli** [comment]
Ah, that's a good point - it would be reproducible in the current Java frontend because we've computed the lattice already, but it wouldn't be in the Python reimplementation. So maybe we should just rename `overload(_)` to something else to avoid the collision?

**Comment by tothtamas28** [comment]
> So maybe we should just rename `overload(_)` to something else to avoid the collision?

I think `overload(_)` is a great name for this feature, and since it's user facing, naming it well is more important than the legacy backend attribute. So I think we should rename the backend attribute.

Also, in general, in order to avoid confusion, we should try to maintain uniqueness of attribute names. (There are a few cases currently that violate this, e.g. `RIGHT` / `RIGHT_INTERNAL`). Order of importance:

```
used in K input > used in KORE output > internal
```

**Comment by Scott-Guest** [comment]
To clarify, the changes I'm making don't strip out any attributes *before* `ModuleToKORE`, they just prevent those attributes from being emitted *during* `ModuleToKORE`, so I don't think the naming conflict would have caused any issues.

Philosophically though, I agree we should keep attribute names unique. If we can get the existing examples cleaned up, it'd be easy to have the whitelist enforce this.

**Comment by Baltoli** [comment]
Ah, I had a slight conceptual misunderstanding about the point at which the stripping would happen. I do agree we should keep them unique though, so it's nice this conversation has sort of pushed us to do it anyway

File: docs/user_manual.md
Comment (Scott-Guest) [inline docs/user_manual.md]: I don't quite follow this example - the `syntax  Val ::= "<" AValType ">" Number` constructor is indeed overly-permissive, but that is always true regardless of any overloading.

Even with a total order, you can still construct similarly problematic examples:
```
module NUMBERS-SYNTAX
    imports INT-SYNTAX
    imports FLOAT-SYNTAX
    syntax SubValType ::= "sub"
    syntax SuperValType ::= "super" | SubValType

    syntax Sub ::= "subVal"
    syntax Super ::= "superVal" | Sub

    syntax SubVal ::= "<" SubValType ">" Sub       [overload(<_>_)]
    syntax SuperVal ::= "<" SuperValType ">" Super [overload(<_>_)] | SubVal
endmodule

module NUMBERS
    imports NUMBERS-SYNTAX
    syntax SuperVal ::= creature32 ( Super ) [function, total]
    rule creature32(S:Sub) => <super> S
    rule creature32(S::Super) => <sub> S
endmodule
```

To me, the over-permissiveness of `syntax  Val ::= "<" AValType ">" Number` is the whole motivation behind overloading in the first place - we want more precise sorting invariants than what K's syntax declaration are able to express.

Still choosing to include the overly-permissive constructor even after using overloading just seems like an odd design decision in the example rather than something fundamental we need to explain to users about overloading. 

Reading through runtimeverification/hs-backend-booster#492, it was mentioned that the problematic constructor was useful because it allows us to write unambiguous rules which handle all overloads like
```
syntax Bar ::= bar(Number) [function, total]
               |  foo(Val) [function, total]
rule foo(< _ > N) => bar(N) // ambiguous without the problematic constructor
```
but I think a more principled and only slightly-inconvenient solution would be to just manually define eliminators
```
syntax AValType ::= valType(Val) [function, total]
rule valType(< V:IValType > _) => V
rule valType(< V:FValType> _ ) => V

syntax Number ::= number(Val) [function, total]
rule number(< _ > N:Int) => N
rule number(< _ > N:Float) => N
```
and write
```
rule foo(V) => bar(number(V))
```

Comment (Baltoli) [reply docs/user_manual.md]: Your analysis all makes sense to me; I guess I read Jost's writeup as a cautionary example that needed to be explained to users, and without thinking it through properly made the incorrect inference that because a non-total overload ordering is prone to producing this kind of broken case, total orderings cannot. I guess we can just cut this section from the manual entirely?

File: docs/user_manual.md
Comment (Scott-Guest) [inline docs/user_manual.md]: This `[^overload-issue]` footnote is dead code now (and while we're at it, `[^1]` is dead too).

File: docs/user_manual.md
Comment (Scott-Guest) [inline docs/user_manual.md]: I don't follow this point - wouldn't substitutability fail to hold because the overload ordering is covariant rather than contravariant in the parameter sorts?

Perhaps a better phrasing is that a more precise overload is a "restriction" of a less precise overload - when the inputs are restricted to a more precise sort, we can also give the output a more precise sort.

File: docs/user_manual.md
Comment (Scott-Guest) [inline docs/user_manual.md]: This isn't a lexicographic ordering - overloading requires that **every** component of `P`'s tuple is a subsort of the corresponding component of `Q`'s tuple, but lexicographic ordering only requires this for the first component where the tuples disagree.

Formally, renaming `S_P` to `S_p0` for ease of notation, the overload ordering defines `P < Q` if
- `S_pi <: S_qi` for all `0 ≤ i ≤ N`
- `(S_p0, S_p1, ..., S_pN) != (S_q0, S_q1, ..., S_qN)`

However, a lexicographic ordering would only require
- There exists some `0 ≤ k ≤ N` such that `S_pi = S_qi` for all `0 ≤ i < k` and `S_pk < S_qk`

**Review by Scott-Guest — APPROVED** [review]
Code LGTM, but a lingering conceptual question - does overloading have any semantic effect or does it only affect the disambiguation process?

For example, in
```
module NUMBERS-SYNTAX
    imports INT-SYNTAX
    imports FLOAT-SYNTAX
    syntax IValType ::= "i32" | "i64"
    syntax FValType ::= "f32" | "f64"
    syntax IVal ::= "<" IValType ">" Int   [overload(<_>_)]
    syntax FVal ::= "<" FValType ">" Float [overload(<_>_)]
endmodule
```
the two overloads are incomparable and thus will not affect disambiguation. Are the `overload` attributes here dead code then (and we should beef up our warning accordingly), or do they have an observable effect?

It would be good to clarify this in the user manual.

File: docs/user_manual.md
Comment (Baltoli) [reply docs/user_manual.md]: Good spot!

File: docs/user_manual.md
Comment (Baltoli) [reply docs/user_manual.md]: Yep, you're right

**Comment by Baltoli** [comment]
> Code LGTM, but a lingering conceptual question - does overloading have any semantic effect or does it only affect the disambiguation process?

I believe it affects [unification](https://github.com/runtimeverification/haskell-backend/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md#L4) in the old Haskell backend, but figuring that algorithm + implementation combo out properly is probably beyond the scope of our current work on this and we'll need to defer to @jberthold et. al. to figure out what should happen here in the future.

**Pull request merged**
